File: code/Java/Chapter5/5-2/Deque.java
import java.util.LinkedList;

public class Deque<T> {
    private LinkedList<T> data;

    public Deque() {
        data = new LinkedList<T>();
    }

    public int size() {
        return data.size();
    }

    public boolean isEmpty() {
        return data.isEmpty();
    }

    public void pushFront(T elem) {
        data.addFirst(elem);
    }

    public void pushBack(T elem) {
        data.addLast(elem);
    }

    public T popFront() {
        return data.removeFirst();
    }

    public T popBack() {
        return data.removeLast();
    }

    public T front() {
        return data.getFirst();
    }

    public T back() {
        return data.getLast();
    }
}

File: code/Python/Chapter5/5-2/deque.py
class Deque:
    def __init__(self):
        self.__data = []
    
    def is_empty(self):
        return self.__data == []
    
    def __len__(self):
        return len(self.__data)
    
    def clear(self):
        self.__data = []
    
    def push_front(self, elem):
        self.__data.insert(0, elem)
    
    def push_back(self, elem):
        self.__data.append(elem)
    
    def pop_front(self):
        return self.__data.pop(0)
    
    def pop_back(self):
        return self.__data.pop()
    
    def front(self):
        return self.__data[0]
    
    def back(self):
        return self.__data[-1]